### PR TITLE
Python3.6.1 support

### PR DESCRIPTION
**Pull request by jmphilli**
This adds a new target lib during compilation which includes the python 3.6 libraries. There were some symbols added which broke pyflame for python 3.6. See [here](https://github.com/uber/pyflame/issues/69)

Tried it out against 3.6.1, 3.4, and 2.7.

### CODE REVIEW

**Comment by jmphilli** [comment]
hmm. having a little trouble getting both py3 versions to play nice. fiddling around.

File: src/symbol.cc [hunk 1]
```diff
@@ -146,8 +146,11 @@ void ELF::WalkTable(int sym, int str, bool &have_version, PyVersion *version,
         *version = PyVersion::Py2;
       } else if (strcmp(name, "PyBytes_Type") == 0) {
         // if we find PyBytes_Type, it's python 3
-        have_version = true;
+        // continue looping though, in case we see a python3.6 symbol
         *version = PyVersion::Py3;
+      } else if (strcmp(name, "_PyEval_RequestCodeExtraIndex") == 0 || strcmp(name, "_PyCode_GetExtra")  == 0|| strcmp(name, "_PyCode_SetExtra") == 0) {
```
Comment (lita) [inline src/symbol.cc]: nit: you need a space between ```strcmp(name, "_PyCode_GetExtra")  == 0|| strcmp(name, "_PyCode_SetExtra") == 0```

Comment (lita) [inline src/symbol.cc]: would be nice to add a comment here

File: src/frob.cc [hunk 2]
```diff
@@ -63,10 +63,22 @@ unsigned long StringData(unsigned long addr) {
...
+
+unsigned long StringData(unsigned long addr) {
+  // this works only if the filename is all ascii *fingers crossed*
+  return addr + sizeof(PyASCIIObject);
+}
 #else
 static_assert(false, "uh oh, bad PYFLAME_PY_VERSION");
 #endif
 
+
```
Comment (jmphilli) [inline src/frob.cc]: undo this

File: src/pyfrob.cc [hunk 3]
```diff
@@ -113,12 +113,18 @@ FROB_FUNCS
 }
 #endif
 
-#ifdef ENABLE_PY3
+#if defined(ENABLE_PY3)
```
Comment (jmphilli) [inline src/pyfrob.cc]: undo this

File: src/pyfrob.cc [hunk 4]
```diff
@@ -141,11 +147,16 @@ void PyFrob::SetPython(PyVersion version) {
       get_threads_ = py2::GetThreads;
       break;
 #endif
-#ifdef ENABLE_PY3
+#if defined(ENABLE_PY3)
```
Comment (jmphilli) [inline src/pyfrob.cc]: undo this

File: src/pyfrob.cc [hunk 5]
```diff
@@ -141,11 +147,16 @@ void PyFrob::SetPython(PyVersion version) {
...
     case PyVersion::Py3:
       get_threads_ = py3::GetThreads;
       break;
 #endif
+#if defined(ENABLE_PY36)
```
Comment (jmphilli) [inline src/pyfrob.cc]: switch to ifdef

File: .travis.yml [hunk 6]
```diff
@@ -18,17 +18,22 @@ matrix:
...
       dist: trusty
       addons:
         apt:
           packages:
-            - python3-dev
+            - python3
       env: PYTHONVERSION=python3
 
+before_install:
+  - sudo add-apt-repository ppa:jonathonf/python-3.6 -y
```
Comment (jmphilli) [inline .travis.yml]: ubuntu trusty 14.04 doesn't have an official ppa for python3.6 so i needed to do this to get the dev artifacts

File: src/pyflame.cc [hunk 7]
```diff
@@ -36,7 +36,7 @@
 #include "./version.h"
 
 // FIXME: this logic should be moved to configure.ac
-#if !defined(ENABLE_PY2) && !defined(ENABLE_PY3)
+#if !defined(ENABLE_PY2) && !defined(ENABLE_PY3) && !defined(ENAABLE_PY36)
```
Comment (eklitzke) [inline src/pyflame.cc]: typo here

File: src/frob.cc [hunk 8]
```diff
@@ -28,7 +28,7 @@
 #include <Python.h>
 #include <frameobject.h>
 
-#if PYFLAME_PY_VERSION == 3
+#if PYFLAME_PY_VERSION == 3 || PYFLAME_PY_VERSION == 36
```
Comment (eklitzke) [inline src/frob.cc]: let's set change the existing references from `PYFLAME_PY_VERSION = 2` to `PYFLAME_PY_VERSION = 20`, and `PYFLAME_PY_VERSION = 3` to `PYFLAME_PY_VERSION = 30`. That way this `#if` can become:

```c
#if PYFLAME_PY_VERSION >= 30
```

File: src/frob.cc [hunk 9]
```diff
@@ -63,6 +63,16 @@ unsigned long StringData(unsigned long addr) {
   // this works only if the filename is all ascii *fingers crossed*
   return addr + sizeof(PyASCIIObject);
 }
+#elif PYFLAME_PY_VERSION == 36
+namespace py36 {
```
Comment (eklitzke) [inline src/frob.cc]: Does this actually need to be a new namespace? I think you could keep the `PYFLAME_PY_VERSION` guard, but reuse the existing `py3` namespace. Then you don't have to change `src/pyfrob.cc`.

**Comment by eklitzke** [comment]
What Linux distro are you using? I'm using Python 3.6.1 on Fedora and Pyflame master works for me. But some distros compile Python differently. I'm interested in seeing if I can repro the issue.

**Comment by jmphilli** [comment]
hmmm weird. i'm on ubuntu trusty 14.04. [this](https://github.com/uber/pyflame/issues/69) is the issue i'm seeing

**Comment by jmphilli** [comment]
i think you should just be able to update the `.travis.yml` file like i did to repro...

**Comment by eklitzke** [comment]
Can you tell me if #72 resolves the issue for you? The code for that PR is in the branch `python3.6`.

**Comment by eklitzke** [comment]
See https://github.com/uber/pyflame/issues/69#issuecomment-314697390

**Comment by eklitzke** [comment]
This branch seems to work on Fedora, so things are simpler than I had thought. However, it still needs some work:

* I redid how the Travis stuff works on master, to fix the 3.4 and 3.5 tests. What you want to do is update this branch to have the `.travis.yml` from master, and the only change you should need is to add 3.6 to the Python version list. No need for the PPA.
* You should be able to undo the `configure.ac` and `Makefile.am` changes you had before, since I believe the non-PPA Python 3.6 should just declare itself as PY3.

Since you will no longer see `ENABLE_PY36` from automake, the code that detects Python 3 vs 3.6 is going to have to change slightly. It will work approximately like this:

* Remove the `py36` namespace
* Keep the existing code in `symbol.cc` that detects the Python version
* `FROB_FUNCS` will have to be updated for Python 3.6 so that it can switch between the two cases *at runtime*

I think the easiest way to do the `FROB_FUNCS` change would be to update the methods `StringSize()` etc. to take a new `PyVersion` parameter. Then you'll pass the `PyVersion` detected from the symbol code down to `StringSize()` when it's called, and that method can dynamically dispatch between a method named something like `StringSize_Py30()` for Python 3.0-3.5, and `StringSize_Py36()` for Python 3.6.

**Comment by jmphilli** [comment]
got it -- thank you @eklitzke -- i'll make the changes

**Comment by jmphilli** [comment]
continuing over here https://github.com/uber/pyflame/pull/81/files

**Pull request closed without merging**